shark_turbine/kernel/wave/expansion.py

```diff
@@ -120,6 +121,50 @@ def is_mma(node: fx.Node) -> bool:
     return mapping
 
 
+def set_node_indices(
+    trace: CapturedTrace,
+    constraints: Sequence[Constraint],
+    mma_index: dict[IndexSymbol, int],
+):
+    """
+    Set the indices of the nodes based on the user constraints. In certain
+    operators (like read, write), there is only a single index associated
+    with the node (the index to read from, the index to write to). But for
+    other operators like mma, each operand reads from a different index.
+
+    Rather than maintain operand specific indices for operators, we maintain
+    dimension specific indices for each operator. So for an mma operator that
+    has a signature of (MxK, NxK) -> MxN, we maintain only 3 mappings for
+    dimensions M, N and K, but allow each mapping to be piecewise conditioned
+    on the operand.
+    """
+
+    def compute_index(node: fx.Node) -> bool:
+        custom = get_custom(node)
+        custom.index = {}
+        for dim in custom.indexing_dims:
+            for constraint in constraints:
+                if (
+                    not isinstance(constraint, HardwareConstraint)
+                    and dim != constraint.dim
+                ):
+                    continue
+                if not custom.index:
+                    custom.index = {
+                        dim: IndexSequence(0, 1) for dim in custom.indexing_dims
```

Comment:
Do we expect the handling of the not custom index and loop over `custom.indexing_dims` in this for loop? would it not make sense to handle this outside of `for dim in custom.indexing_dims:`?

Reply:
My intention here was to only set the index attribute on operators that were affected by the constraints. The reason for this is to distinguish between the following 3 scenarios: an operator that has no index, an operator that has an index but the index is None and an operator that has an index that is not None. I dont think we need to distinguish between the first 2 options - so we can always set the index, but the index could be None. So with that I could move the initialization of custom_index outside the loop.